### ISSUE TITLE
[BUG] WORKDIR creates extra layer

### ISSUE THREAD

**Issue by MartinHotmann** [docker/buildx#1910]
### Contributing guidelines

- [X] I've read the [contributing guidelines](https://github.com/docker/buildx/blob/master/.github/CONTRIBUTING.md) and wholeheartedly agree

### I've found a bug and checked that ...

- [ ] ... the documentation does not mention anything about my problem
- [ ] ... there are no open or closed issues that are related to my problem

### Description

I previously have created a lot of containers, but now, since some time. When I use `WORKDIR` it will create a new layer.
This goes against a lot of reports and some other readings I had. I don't know if this 'now' needs to be like this.

Let me elaborate:

I use a multistage DOCKERFILE which starts with `golang:1-alpine`. Here the shorted version:

```
##
## SCHRITT 1 - BAUEN
##

FROM golang:1-alpine as build

WORKDIR /app/public

COPY ./PUSH /app

ARG TARGETOS TARGETARCH

RUN some code to execute here, which creeates a binary.


##
## SCHRITT 2 - KOPIEREN & AUSFÜHREN
##

FROM scratch as run

COPY --from=build /ALL_TEMP /

WORKDIR /app/public
EXPOSE 8003
ENTRYPOINT ["/app/public/app"]
```

When I remove the second `WORKDIR /app/public` it will result in 1 layer.
If I leave it, it will be 2 layers.
The second layer is 32b big and contains just the change of the working directory.
![image](https://github.com/docker/buildx/assets/17692860/95877934-005d-46b6-8bef-07fa842055f5)

It actually should be a 0b layer (like every other, that is not `ADD`, `RUN` or `COPY`)

This was not the case, some months/weeks ago.

I am building on Debian bullseye, my setup is:

- Docker: `v24.0.2, build cb74dfc`
- Docker compose: `v2.19.0`
- Buildx: `v0.10.5 86bdced`

If you need any additional info please let me know.

I also read, that `WORKDIR` will be inharited within a multistep BUILD, but this is just true for the build process, once the app ist started, and `WORKDIR` is missing, the workingdirectory is `/` and not the `WORKDIR` set in a previous step of the build.

### Expected behaviour

`WORKDIR` should not be a layer.

### Actual behaviour

When using `WORKDIR` in a docker scratch build it will create a seperate layer.

### Buildx version

`v0.10.5 86bdced`

### Docker info

```text
Client: Docker Engine - Community
 Version:    24.0.2
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.5
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.18.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose

Server:
 Containers: 6
  Running: 6
  Paused: 0
  Stopped: 0
 Images: 201
 Server Version: 24.0.2
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 3dce8eb055cbb6872793272b4f20ed16117344f8
 runc version: v1.1.7-0-g860f061
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.10.0-22-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 10
 Total Memory: 31.35GiB
 Name: clicon.hosting-hotmann.de
 ID: ZUV6:WTSH:O2CJ:ZU5E:OPCZ:2K6B:X4BB:BAQS:LY23:ADAG:JPIL:EOHT
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: h0tmann
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```
```


### Builders list

```text
NAME/NODE    DRIVER/ENDPOINT             STATUS  BUILDKIT                              PLATFORMS
multiarch *  docker-container
  multiarch0 unix:///var/run/docker.sock running v0.11.6                               linux/amd64*, linux/amd64/v2*, linux/amd64/v3*, linux/386*, linux/arm64*, linux/arm/v8*, linux/arm/v7*, linux/arm/v6*, linux/riscv64*, linux/ppc64le*, linux/s390x*, linux/mips64le, linux/mips64, linux/ppc64
default      docker
  default    default                     running v0.11.7-0.20230525183624-798ad6b0ce9f linux/amd64, linux/amd64/v2, linux/amd64/v3, linux/386, linux/arm64, linux/riscv64, linux/ppc64, linux/ppc64le, linux/s390x, linux/mips64le, linux/mips64, linux/arm/v7, linux/arm/v6
```

```
##
## SCHRITT 1 - BAUEN
##

FROM golang:1-alpine as build

WORKDIR /app/public

COPY ./PUSH /app

ARG TARGETOS TARGETARCH

RUN some code to execute here, which creeates a binary.


##
## SCHRITT 2 - KOPIEREN & AUSFÜHREN
##

FROM scratch as run

COPY --from=build /ALL_TEMP /

WORKDIR /app/public
EXPOSE 8003
ENTRYPOINT ["/app/public/app"]
```

tl;dr:

`WORKDIR` creates an extra layer, while just `COPY`, `RUN` and `ADD` should ever create layers - `WORKDIR` should contain `0b` like `EXPOSE` or any of the other informational commands.

**Comment by jedevc** [comment]
This behavior is intentional. See https://github.com/moby/buildkit/blob/master/frontend/dockerfile/docs/reference.md#workdir (emphasis mine):

> The WORKDIR instruction sets the working directory for any RUN, CMD, ENTRYPOINT, COPY and ADD instructions that follow it in the Dockerfile. **If the WORKDIR doesn't exist, it will be created even if it's not used in any subsequent Dockerfile instruction.**

This is where the layer is coming from, it's creating the directory. Even if the directory already exists, we still need this layer.
